src/rpc/blockchain.cpp

```diff
@@ -1501,14 +1502,82 @@ UniValue MempoolInfoToJSON(const CTxMemPool& pool)
     ret.pushKV("unbroadcastcount", uint64_t{pool.GetUnbroadcastTxs().size()});
+
+    if (feeLimits) {
+        const MempoolHistogramFeeLimits& limits = feeLimits.value();
+
+        /* Keep histogram per...
+         * ... cumulated tx sizes
+         * ... txns (count)
+         * ... cumulated fees */
+        std::vector<uint64_t> sizes(limits.size(), 0);
+        std::vector<uint64_t> count(limits.size(), 0);
+        std::vector<uint64_t> fees(limits.size(), 0);
+
+        for (const CTxMemPoolEntry& e : pool.mapTx) {
+            const int size = (int)e.GetTxSize();
+            CAmount fee = e.GetFee();
+            uint64_t asize = e.GetSizeWithAncestors();
+            CAmount afees = e.GetModFeesWithAncestors();
+            uint64_t dsize = e.GetSizeWithDescendants();
+            CAmount dfees = e.GetModFeesWithDescendants();
+
+            CAmount fpb = fee / size; // fee per byte
+            CAmount afpb = afees / asize; // fee per byte including ancestors
+            CAmount dfpb = dfees / dsize; // fee per byte including descendants
+            CAmount tfpb = (afees + dfees - fee) / (asize + dsize - size);
+            CAmount feeperbyte = std::max(std::min(dfpb, tfpb), std::min(fpb, afpb));
+
+            // Distribute feerates into feelimits
+            for (int i = limits.size() - 1; i >= 0; --i) {
+                if (feeperbyte >= limits[i].second) {
+                    sizes[i] += size;
+                    ++count[i];
+                    fees[i] += fee;
+                    break;
+                }
+            }
+        }
+
+        CAmount total_fees = 0; // Track total amount of available fees in mempool
+        UniValue ranges(UniValue::VOBJ);
+        for (size_t i = 0; i < limits.size(); ++i) {
+            UniValue info_sub(UniValue::VOBJ);
+            info_sub.pushKV("sizes", sizes[i]);
+            info_sub.pushKV("count", count[i]);
+            info_sub.pushKV("fees", fees[i]);
+            info_sub.pushKV("from_feerate", limits[i].first);
+
+            if (i == limits.size() - 1) {
+                info_sub.pushKV("to_feerate", "Max"); // TODO.
+            } else {
+                info_sub.pushKV("to_feerate", limits[i + 1].first);
+            }
+
+            total_fees += fees[i];
+            ranges.pushKV(limits[i].first, info_sub);
+        }
+
+        UniValue info(UniValue::VOBJ);
+        info.pushKV("ranges", ranges);
+        info.pushKV("total_fees", total_fees);
+        ret.pushKV("fee_histogram", info);
+    }
+
     return ret;
 }
 
 static RPCHelpMan getmempoolinfo()
 {
     return RPCHelpMan{"getmempoolinfo",
                 "\nReturns details on the active state of the TX memory pool.\n",
-                {},
+                {
+                    {"fee_histogram", RPCArg::Type::ARR, RPCArg::Optional::OMITTED_NAMED_ARG, "Fee amounts",
+                        {
+                            {"fee", RPCArg::Type::AMOUNT, RPCArg::Optional::NO, "A fee amount"},
```

Comment:
IIUC these should be fee rates, not fees
```suggestion
                            {"fee_rate", RPCArg::Type::AMOUNT, RPCArg::Optional::NO, "Fee rate to group the fees by"},
```